docs: updates to micrometer docs (#1539)
elastic/apm-agent-java: docs/metrics.asciidoc

--

[[metrics-micrometer]]
=== Micrometer metrics

> beta::[]

Author: The `beta::[]` tag builds a beta admonition paragraph:

<img width="768" alt="Screen Shot 2021-05-04 at 8 46 11 AM" src="https://user-images.githubusercontent.com/5618806/117031212-35680680-acb5-11eb-8ddb-34754d0ecd79.png">

> beta::[]

Author: Should be available now: https://apm-agent-java_1539.docs-preview.app.elstc.co/guide/en/apm/agent/java/master/metrics-micrometer.html

you'll only be storing one out of 10 requests.
The labels you set on non-sampled transactions will be lost.

[float]
[[metrics-micrometer-beta-caveats]]
==== Caveats

To fix some of the caveats listed here, we may have to introduce breaking changes.
Please look for Micrometer-related release notes before updating.

* Metrics with dots in their name can result in mapping conflicts.
See https://github.com/elastic/apm/issues/347[elastic/apm#347]

> * Histograms aren't supported yet

Reviewer: It's not completely accurate that Histograms are not supported. Histograms are partially supported creating two counters that already enable insightful information: a sum counter (xxx.sum) and a count counter (xxx.count).

Instead of the explanation "Histograms aren’t supported yet", an explanation that would look like

> Histograms ([`DistributionSummary`](https://www.javadoc.io/doc/io.micrometer/micrometer-core/latest/io/micrometer/core/instrument/DistributionSummary.html), [`Timer`](https://www.javadoc.io/doc/io.micrometer/micrometer-core/latest/io/micrometer/core/instrument/Timer.html)and [`LongTaskTimer`](https://www.javadoc.io/doc/io.micrometer/micrometer-core/latest/io/micrometer/core/instrument/LongTaskTimer.html)) are partially supported converting the histogram metric into two derived metric: a counter of the values and the sum of the values.
> For example, `DistributionSummary.builder("order").register(...).record(orderPrice)` will create two metrics: `order.sum`and `order.count`.
 
TODO: we should verify that `Timer` and `LongTaskTimer` behave the same way as `DistributionSummary` creating these two derived metrics.

you'll only be storing one out of 10 requests.
The labels you set on non-sampled transactions will be lost.

[float]
[[metrics-micrometer-beta-caveats]]
==== Caveats

To fix some of the caveats listed here, we may have to introduce breaking changes.
Please look for Micrometer-related release notes before updating.

* Metrics with dots in their name can result in mapping conflicts.

> See https://github.com/elastic/apm/issues/347[elastic/apm#347]

Reviewer 2: Much better 🙂

  But you have to remember to group by a combination of dimensions that uniquely identify the time series.
  This may be a combination of `host.name` and `service.name`, or the `kubernetes.pod.id`.

[float]

> [[micrometer-get-started]]
> ==== Get started with Micrometer

Reviewer 2: ```suggestion
[[metrics-micrometer-get-started-existing]]
==== Get started with existing Micrometer setup
```

Reviewer 2: Assuming that the duplicated one is removed from below

You should now see documents containing both metrics collected by the APM agent and custom metrics from micrometer.
Narrow your search with a known micrometer metric field.
For example, if you know `cache.puts` is a micrometer field, add `cache.puts: *` to your search to return only micrometer metrics documents:

[float]
[[visualize-micrometer-data]]
==== Visualize micrometer data

NOTE: Monotonically increased counters and Positive rate aggregations are not fully supported in the current version.

{kib-ref}/tsvb.html[TSVB] is the recommended visualization for micrometer metrics.
First, make sure to select right aggregation. The most common options are:

* Sum: Useful for business metrics
* Average: Usually used for performance related metrics

It's a common task to group metrics by attributes, including micrometer labels or attributes already collected by APM agents. This could be service versions, runtime versions, or even cloud metadata.

TIP: See the https://www.elastic.co/blog/visualizing-observability-with-kibana-event-rates-and-rate-of-change-in-tsvb[Event rates and rate of change in TSVB] blog post for more information.

[float]

> [[metrics-micrometer-get-started-existing]]

Reviewer 2: Looks like a duplication

Author: Done ✅

Author: ✅

If you are using multiple registries within a `CompoundMeterRegistry`, the agent only reports the metrics once.

[float]
[[verify-micrometer-data]]
==== Verify Micrometer data

Use Discover to validate that metrics are successfully reported to Kibana.

. Launch {kib}.
. Open the main menu, then click *Discover*.
. Select `apm-*` as your index pattern.
. Filter the data to only show documents with metrics: `processor.name :"metric"`
. Optionally, apply additional filters by service or host names if Micrometer was only instrumented on a subset of your environment.

You should now see documents containing both metrics collected by the APM agent and custom metrics from Micrometer.
Narrow your search with a known Micrometer metric field.
For example, if you know you have registered the metric name `cache.puts` in the Micrometer `MeterRegistry`, add `cache_puts: *` (dots are replaced with underscores) to your search to return only Micrometer metrics documents:

[float]
[[visualize-micrometer-data]]
==== Visualize Micrometer data

NOTE: Monotonically increased counters and Positive rate aggregations are not fully supported in the current version.

> {kib-ref}/tsvb.html[TSVB] is the recommended visualization for Micrometer metrics.

Author: ```suggestion
{kibana-ref}/tsvb.html[TSVB] is the recommended visualization for Micrometer metrics.
```

            return null;
        }
    }, Clock.SYSTEM);
----

> When using Spring Boot, you can use the `management.metrics.export.simple` prefix to configure via `application.properties`

Reviewer: I didn't see the Spring Boot section when I used these docs to setup my micrometer integration.
Did we consider solutions to make it easy for developers using Spring Boot to find our dedicated doc. As Spring Boot is very popular, we may even consider moving this section before the "Do It Yourself" section.

Reviewer 3: Let's add a separate section for Spring Boot and add a table of contents to the top of the page.

```suggestion
==== Get started with Spring Boot

The easiest way to get started with Spring Boot is to add a dependency to [Spring Boot Actuator](https://docs.spring.io/spring-boot/docs/current/reference/html/production-ready-features.html) if you haven't already. Spring Boot Actuator provides dependency management and auto-configuration for Micrometer.

Use the `management.metrics.export.simple` prefix to configure via `application.properties`
```
